pkcs5/src/pbes2/encryption.rs

```diff
@@ -49,61 +49,55 @@ fn cbc_decrypt<'a, C: BlockCipherDecrypt + BlockCipher + KeyInit>(
         .map_err(|_| Error::EncryptFailed)
 }
 
-fn gcm_encrypt<'a, C>(
+fn gcm_encrypt<'a, C, NonceSize, TagSize>(
     es: EncryptionScheme,
     key: EncryptionKey,
-    iv: &[u8],
+    nonce: Nonce<NonceSize>,
     buffer: &'a mut [u8],
     pos: usize,
 ) -> Result<&'a [u8]>
 where
     C: BlockCipher + BlockSizeUser<BlockSize = U16> + GcmKeyInit + BlockCipherEncrypt,
-    aes_gcm::AesGcm<C, U12>: GcmKeyInit,
+    aes_gcm::AesGcm<C, NonceSize, TagSize>: GcmKeyInit,
+    TagSize: aes_gcm::TagSize,
+    NonceSize: aes::cipher::ArraySize,
 {
-    const TAG_SIZE: usize = 16;
-    const NONCE_SIZE: usize = 12;
-
-    if buffer.len() < TAG_SIZE + pos {
+    if buffer.len() < TagSize::USIZE + pos {
         return Err(Error::EncryptFailed);
     }
-    let gcm = <aes_gcm::AesGcm<C, U12> as GcmKeyInit>::new_from_slice(key.as_slice())
-        .map_err(|_| es.to_alg_params_invalid())?;
-    let nonce: [u8; NONCE_SIZE] = iv.try_into().map_err(|_| es.to_alg_params_invalid())?;
-    let nonce = Nonce::from(nonce);
+    let gcm =
+        <aes_gcm::AesGcm<C, NonceSize, TagSize> as GcmKeyInit>::new_from_slice(key.as_slice())
+            .map_err(|_| es.to_alg_params_invalid())?;
     let tag = gcm
         .encrypt_in_place_detached(&nonce, &[], &mut buffer[..pos])
         .map_err(|_| Error::EncryptFailed)?;
     buffer[pos..].copy_from_slice(tag.as_ref());
-    Ok(&buffer[0..pos + TAG_SIZE])
+    Ok(&buffer[0..pos + TagSize::USIZE])
 }
 
-fn gcm_decrypt<'a, C>(
+fn gcm_decrypt<'a, C, NonceSize, TagSize>(
     es: EncryptionScheme,
     key: EncryptionKey,
-    iv: &[u8],
+    nonce: Nonce<NonceSize>,
     buffer: &'a mut [u8],
 ) -> Result<&'a [u8]>
 where
     C: BlockCipher + BlockSizeUser<BlockSize = U16> + GcmKeyInit + BlockCipherEncrypt,
     aes_gcm::AesGcm<C, U12>: GcmKeyInit,
+    TagSize: aes_gcm::TagSize,
+    NonceSize: aes::cipher::ArraySize,
 {
-    const TAG_SIZE: usize = 16;
-    const NONCE_SIZE: usize = 12;
-
-    let msg_len = if buffer.len() < TAG_SIZE {
+    let msg_len = if buffer.len() < TagSize::USIZE {
         return Err(Error::DecryptFailed);
     } else {
-        buffer.len() - TAG_SIZE
+        buffer.len() - TagSize::USIZE
     };
 
-    let gcm = <aes_gcm::AesGcm<C, U12> as GcmKeyInit>::new_from_slice(key.as_slice())
+    let gcm = <aes_gcm::AesGcm<C, NonceSize> as GcmKeyInit>::new_from_slice(key.as_slice())
```

Comment:
```suggestion
    let gcm = <aes_gcm::AesGcm<C, NonceSize, TagSize> as GcmKeyInit>::new_from_slice(key.as_slice())
```

This will otherwise take the default TagSize (U16)